File: frontend/src/validation/form_data.ts
// import * as Yup from 'yup';

// function buildYupSchema(inputs) {
//   const shape = {};

//   inputs.forEach((input) => {
//     let schema;

//     // Start with base type
//     switch (input.type) {
//       case 'email':
//       case 'text':
//       case 'textarea':
//         schema = Yup.string();
//         break;
//       case 'number':
//         schema = Yup.number();
//         break;
//       case 'date':
//         schema = Yup.date();
//         break;
//       case 'checkbox':
//         schema = Yup.array();
//         break;
//       case 'select':
//       case 'radio':
//         schema = Yup.string();
//         break;
//       default:
//         schema = Yup.string();
//     }

//     // Apply required
//     if (input.required) {
//       schema = schema.required();
//     }

//     // Apply validations if they exist
//     if (input.validation) {
//       const v = input.validation;

//       // String validations
//       if (v.minLength) schema = schema.min(v.minLength);
//       if (v.maxLength) schema = schema.max(v.maxLength);
//       if (v.length) schema = schema.length(v.length);
//       if (v.matches) schema = schema.matches(new RegExp(v.matches));
//       if (v.email) schema = schema.email();
//       if (v.url) schema = schema.url();
//       if (v.uuid) schema = schema.uuid();

//       // Number validations
//       if (v.min !== undefined) schema = schema.min(v.min);
//       if (v.max !== undefined) schema = schema.max(v.max);
//       if (v.lessThan !== undefined) schema = schema.lessThan(v.lessThan);
//       if (v.moreThan !== undefined) schema = schema.moreThan(v.moreThan);
//       if (v.positive) schema = schema.positive();
//       if (v.negative) schema = schema.negative();
//       if (v.integer) schema = schema.integer();

//       // Array validations
//       if (v.minItems) schema = schema.min(v.minItems);
//       if (v.maxItems) schema = schema.max(v.maxItems);

//       // Date validations
//       if (v.minDate) schema = schema.min(new Date(v.minDate));
//       if (v.maxDate) schema = schema.max(new Date(v.maxDate));
//     }

//     shape[input.name] = schema;
//   });

//   return Yup.object().shape(shape);
// }

// // Usage
// const formData = await fetchForm(formId);
// const validationSchema = buildYupSchema(
//   formData.steps.flatMap((s) => s.inputs)
// );

// function evaluateOperator(
//   fieldValue: any,
//   operator: CondOperator,
//   targetValue: any
// ): boolean {
//   switch (operator) {
//     case 'equals':
//       return fieldValue === targetValue;
//     case 'not_equals':
//       return fieldValue !== targetValue;
//     case 'contains':
//       return fieldValue?.includes(targetValue);
//     case 'greater_than':
//       return fieldValue > targetValue;
//     case 'less_than':
//       return fieldValue < targetValue;
//     case 'is_empty':
//       return !fieldValue || fieldValue === '';
//     case 'is_not_empty':
//       return !!fieldValue && fieldValue !== '';
//     case 'is_completed':
//       return fieldValue === true;
//     default:
//       return true;
//   }
// }

File: backend/pkg/repositories/form/form_data.go
package form_repo

type FormData struct {
	Steps []Step `json:"steps"`
	Hero  *Hero  `json:"hero,omitempty"`
}

type Step struct {
	UUID        string     `json:"uuid"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	Fields      []Field    `json:"fields"`
	Condition   *Condition `json:"condition,omitempty"` // Optional condition object
}

type Field struct {
	UUID         string      `json:"uuid"`
	Type         string      `json:"type"` // "text", "email", "textarea", "select", "radio", "checkbox", "number", "date", etc.
	Name         string      `json:"name"`
	Label        string      `json:"label"`
	Placeholder  string      `json:"placeholder,omitempty"`
	Required     bool        `json:"required"`
	Validation   *Validation `json:"validation,omitempty"`
	Options      []Option    `json:"options,omitempty"`
	DefaultValue string      `json:"default_value,omitempty"`
	Condition    *Condition  `json:"condition,omitempty"` // Optional condition object
}

type Condition struct {
	Operator   string     `json:"operator"`   // "AND" or "OR"
	Conditions []CondRule `json:"conditions"` // Array of condition rules
}

type CondRule struct {
	Type     string      `json:"type"`     // "field" or "step"
	UUID     string      `json:"uuid"`     // UUID of the field or step to check
	Operator string      `json:"operator"` // "equals", "notEquals", "contains", "greaterThan", "lessThan", "isEmpty", "isNotEmpty"
	Value    interface{} `json:"value"`    // Value to compare against
}

type Validation struct {
	// String validations
	MinLength *int    `json:"min_length,omitempty"` // .min(n)
	MaxLength *int    `json:"max_length,omitempty"` // .max(n)
	Length    *int    `json:"length,omitempty"`     // .length(n)
	Matches   *string `json:"matches,omitempty"`    // .matches(regex)
	Email     *bool   `json:"email,omitempty"`      // .email()
	URL       *bool   `json:"url,omitempty"`        // .url()
	UUID      *bool   `json:"uuid,omitempty"`       // .uuid()

	// Number validations
	Min      *float64 `json:"min,omitempty"`       // .min(n)
	Max      *float64 `json:"max,omitempty"`       // .max(n)
	LessThan *float64 `json:"less_than,omitempty"` // .lessThan(n)
	MoreThan *float64 `json:"more_than,omitempty"` // .moreThan(n)
	Positive *bool    `json:"positive,omitempty"`  // .positive()
	Negative *bool    `json:"negative,omitempty"`  // .negative()
	Integer  *bool    `json:"integer,omitempty"`   // .integer()

	// Array validations (for multi-select, checkboxes)
	MinItems *int `json:"min_items,omitempty"` // .min(n)
	MaxItems *int `json:"max_items,omitempty"` // .max(n)

	// Date validations
	MinDate *string `json:"min_date,omitempty"` // .min(date) - ISO string
	MaxDate *string `json:"max_date,omitempty"` // .max(date) - ISO string
}

type Option struct {
	UUID  string `json:"uuid"`
	Label string `json:"label"`
	Value string `json:"value"`
}

type Hero struct {
	Src string `json:"src"`
	Alt string `json:"alt"`
}
